pgjdbc/src/main/java/org/postgresql/sasl/ScramAuthenticator.java

```diff
@@ -0,0 +1,154 @@
+  private static final String EXPECTED_SERVER_ADVERTISED_METHODS = "SCRAM-SHA-256";
+
+  private final String user;
+  private final String password;
+  private final PGStream pgStream;
+  private final ScramClient scramClient = ScramClient
+      .channelBinding(ScramClient.ChannelBinding.NO)
+      .stringPreparation(StringPreparations.NO_PREPARATION)
+      .selectClientMechanism(ScramMechanisms.SCRAM_SHA_256)
+      .setup();
+  private ScramSession scramSession;
+  private ScramSession.ServerFirstProcessor serverFirstProcessor;
+  private ScramSession.ClientFinalProcessor clientFinalProcessor;
+
+  private static void log(Level level, String message, Object... args) {
+    if (LOGGER.isLoggable(level)) {
+      LOGGER.log(level, message, args);
+    }
+  }
+
+  private static void logFinest(String message, Object... args) {
+    log(Level.FINEST, message, args);
+  }
+
+  private static void throwConnectionRejectedException(String message, Object... args)
+      throws PSQLException {
+    throw new PSQLException(
+        GT.tr(message, args),
+        PSQLState.CONNECTION_REJECTED
+    );
+  }
+
+  @FunctionalInterface
+  private interface BodySender {
+    void sendBody(PGStream pgStream) throws IOException;
+  }
+
+  private void authenticationMessageSender(BodySender bodySender, int bodyLength)
+      throws IOException {
+    pgStream.sendChar('p');
+    pgStream.sendInteger4(Integer.BYTES + bodyLength);
+    bodySender.sendBody(pgStream);
+    pgStream.flush();
+  }
+
+  public ScramAuthenticator(String user, String password, PGStream pgStream) {
+    this.user = user;
+    this.password = password;
+    this.pgStream = pgStream;
+  }
+
+  public void processServerMechanismsAndInit() throws IOException, PSQLException {
+    String scramAuthenticationMechanism = pgStream.receiveString();
+    pgStream.receiveChar(); // '0' terminating the String
+    if (!EXPECTED_SERVER_ADVERTISED_METHODS.equals(scramAuthenticationMechanism)) {
+      throwConnectionRejectedException(
+          "Unsupported SCRAM mechanism(s) '{0}' advertised by the server",
+          scramAuthenticationMechanism
+      );
+    }
+    logFinest(" Using SCRAM mechanism {0}", EXPECTED_SERVER_ADVERTISED_METHODS);
+
+    scramSession =
+        scramClient.scramSession("*");   // Real username is ignored by server, uses startup one
+  }
+
+  public void sendScramClientFirstMessage() throws IOException {
+    String clientFirstMessage = scramSession.clientFirstMessage();
+    logFinest(" FE=> SASLInitialResponse( {0} )", clientFirstMessage);
+
+    String scramMechanismName = scramClient.getScramMechanism().getName();
+    authenticationMessageSender(
+        s -> {
+          s.send(scramMechanismName.getBytes(StandardCharsets.UTF_8));
+          s.sendChar(0); // List terminated in '\0'
+          s.sendInteger4(clientFirstMessage.length());
+          s.send(clientFirstMessage.getBytes(StandardCharsets.UTF_8));
```

Comment:
This seems to assume `getBytes(StandardCharsets.UTF_8) == clientFirstMessage.length()`.
Is it always the case? If so, please clarify (i.e. in comments) why.

Otherwise I'm inclined to use `byte[].length` instead of `String.length` here (and below)